text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1045 @@
+(as they can contain types that implement `Drop`). In that case the type's drop glue is invoked.
+
+The `Destruct` trait is a bound for whether a type has drop glue. This is trivally true for all types.
+
+`(const) Destruct` trait bounds are satisfied only if the type's `Drop` impl (if any) is `const` and all of the types of
+its components are `(const) Destruct`.
+
+While this means that it's a breaking change to add a type with a non-const `Drop` impl to a type,
+that's already true and nothing new:
+
+```rust
+pub struct S {
+    x: u8,
+    y: Box<()>, // adding this field breaks code.
+}
+
+const fn f(_: S) {}
+//~^ ERROR destructor of `S` cannot be evaluated at compile-time
+```
+
+## Trivially enabled features
+
+You can use `==` operators on most types from libstd from within const contexts.
+
+```rust
+const _: () = {
+    let a = [1, 2, 3];
+    let b = [4, 5, 6];
+    assert!(a != b);
+};
+const _: () = {
+    let a = Some(42);
+    let b = a;
+    assert!(a == b);
+};
+```
+
+Note that the use of `assert_eq!` is waiting on `Debug` impls becoming `const`, which
+will likely be tracked under a separate feature gate under the purview of T-libs.
+Similarly other traits will be made `const` over time, but doing so will be
+unblocked by this feature.
+
+### `const Fn*` traits
+
+All `const fn` implement the corresponding `const Fn()` trait:
+
+```rust
+const fn foo<F: (const) Fn()>(f: F) {
+    f()
+}
+
+const fn bar() {
+    foo(baz)
+}
+
+const fn baz() {}
+```
+
+Arguments and the return type of such functions and bounds follow the same rules as
+their non-const equivalents, so you may have to add `(const)` bounds to other generic
+parameters, too:
+
+
+```rust
+const fn foo<T: (const) Debug, F: (const) Fn(T)>(f: F, arg: T) {
+    f(arg)
+}
+
+const fn bar<T: (const) Debug>(arg: T) {
+    foo(baz, arg)
+}
+
+const fn baz<T: (const) Debug>() {}
+```
+
+For closures and them implementing the `Fn` traits, see the [Future possibilities](#future-possibilities) section.
+
+## Crate authors: Making your own custom types easier to use
+
+You can make trait impls of many standard library traits for your own types have `const` methods.
+While it was often possible to write the same code in inherent methods, operators were
+covered by traits from `std::ops` and thus not avaiable for const contexts.
+Most of the time it suffices to add `const` before the methods in the impl block.
+The compiler will guide you and suggest where to also
+add `(const)` bounds for trait bounds on generic parameters of methods or the impl.
+
+Similarly you can make your traits available for users of your crate to implement constly.
+Note that this has two caveats that are actually the same:
+
+* you cannot mark more methods as `(const)` later,
+* you must decide whether to make a new method `(const)` or not when adding a new method with a default body.
```

Comment:
Could you add an example of a trait with a `(const)`/`~const` method, with a default body that calls other trait methods?

Reply:
For example, the full definition of the `const`ified `PartialEq` trait.